File: worker/src/models/schemas.ts
/**
 * TypeScript types mirroring the Pydantic schemas from web/api/models/schemas.py.
 * These are used for response serialization and internal type safety.
 */

// ── Episode ──

export interface EpisodeSummary {
  broadcast: number
  date: string
  title: string
  track_count: number
  unique_artists: number
  repeat_rate: number | null
}

export interface EpisodeDetail {
  broadcast: number
  date: string
  title: string
  url: string
  tracks: TrackInfo[]
  bandcamp_links: BandcampLink[]
  stats: EpisodeStats | null
}

export interface TrackInfo {
  hour: number
  position: number
  artist: string
  title: string
  album: string | null
}

export interface BandcampLink {
  url: string
  label: string
}

export interface EpisodeStats {
  track_count: number
  unique_artists: number
  repeat_rate: number
}

// ── Artist ──

export interface ArtistSummary {
  artist: string
  plays: number
  episodes: number
  rli: number
  coverage: number | null
  recency_index: number | null
  trend: string | null
  streak: number | null
  album_diversity: number | null
  first_episode: string | null
  last_episode: string | null
  badge: string | null
}

export interface ArtistDetail {
  id: number | null
  artist: string
  plays: number
  episodes: number
  rli: number
  coverage: number | null
  first_appearance: string | null
  last_appearance: string | null
  streak: number | null
  album_count: number | null
  album_diversity: number | null
  top_tracks: TrackCount[]
  album_breakdown: AlbumBreakdown[]
  timeline: TimelinePoint[]
  enrichment: ArtistEnrichment | null
  badge: string | null
}

export interface TrackPlay {
  broadcast: number | null
  date: string | null
}

export interface TrackCount {
  title: string
  plays: number
  last_played: string | null
  last_broadcast: number | null
  broadcasts: TrackPlay[]
  episodes: number | null
  artist: string | null
  album: string | null
}

export interface AlbumBreakdown {
  album: string
  artist: string | null
  plays: number
  distinct_tracks: number
  artwork_url: string | null
  artwork_url_large: string | null
  release_date: string | null
}

export interface TimelinePoint {
  broadcast: number | null
  date: string
  plays: number
}

export interface ArtistEnrichment {
  mbid: string | null
  canonical_name: string | null
  country: string | null
  formed_year: number | null
  genres: string[] | null
  tags: string[] | null
  bio_summary: string | null
  lastfm_tags: string[] | null
  lastfm_bio: string | null
  lastfm_listeners: number | null
  lastfm_playcount: number | null
  lastfm_url: string | null
}

// ── Album ──

export interface AlbumSummary {
  album: string
  artist: string
  plays: number
  distinct_tracks: number
  episodes: number
  last_played: string | null
  artwork_url: string | null
}

export interface ReleaseInfo {
  mbid: string | null
  title: string | null
  status: string | null
  date: string | null
  country: string | null
  format: string | null
  label: string | null
  track_count: number | null
}

export interface AlbumDetail {
  id: number | null
  artist_id: number | null
  album: string
  artist: string
  plays: number
  distinct_tracks: number
  total_tracks: number | null
  episodes: number
  tracks: TrackCount[]
  timeline: TimelinePoint[]
  heatmap: Array<{ year: string; month: string; plays: number }>
  artwork_url: string | null
  artwork_url_large: string | null
  mbid: string | null
  release_group_mbid: string | null
  release_date: string | null
  unplayed_tracks: string[]
  releases: ReleaseInfo[]
}

// ── Recommend / Bandcamp ──

export interface RecommendItem {
  url: string
  bandcamp_artist: string
  album_title: string
  episode_count: number
  first_seen: string | null
  last_seen: string | null
  episodes: number[]
}

export interface RecommendArtist {
  bandcamp_artist: string
  unique_albums: number
  total_episodes: number
  last_seen: string | null
}

// ── Stats ──

export interface OverviewStats {
  episodes: number
  tracks: number
  unique_artists: number
  bandcamp_links: number
  date_range: [string, string] | null
}

export interface TopItem {
  name: string
  value: number
  extra: Record<string, unknown>
}

// ── Admin ──

export interface Cluster {
  base_name: string
  total_tracks: number
  variant_count: number
  variants: Array<{ id: number; name: string; tracks: number; albums: number }>
}

export interface MergePreview {
  preview: boolean
  target_id: number
  target_name: string
  total_affected_tracks: number
  total_affected_albums: number
  details: Array<{
    source_id: number
    source_name: string
    affected_tracks: number
    affected_albums: number
  }>
}

File: web/api/models/schemas.py
"""Pydantic schemas for API responses."""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel


# ── Episode ──

class EpisodeSummary(BaseModel):
    broadcast: int
    date: str
    title: str
    track_count: int
    unique_artists: int
    repeat_rate: float | None = None


class EpisodeDetail(BaseModel):
    broadcast: int
    date: str
    title: str
    url: str
    tracks: list["TrackInfo"]
    bandcamp_links: list["BandcampLink"]
    stats: Optional["EpisodeStats"] = None
    debutants: list["DebutArtist"] = []


class TrackInfo(BaseModel):
    hour: int
    position: int
    artist: str
    title: str
    album: str | None = None
    artist_first: bool = False
    track_first: bool = False


class BandcampLink(BaseModel):
    url: str
    label: str = ""


class EpisodeStats(BaseModel):
    track_count: int
    unique_artists: int
    repeat_rate: float
    debuting_artists: int = 0


class DebutArtist(BaseModel):
    artist: str
    title: str
    album: str | None = None


# ── Artist ──

class ArtistSummary(BaseModel):
    artist: str
    plays: int
    episodes: int
    rli: float  # Bayesian Rollins Love Index (shrinkage toward mean)
    coverage: float | None = None  # % of all episodes this artist appears in
    recency_index: float | None = None
    trend: str | None = None  # "heating_up" | "cooling_down" | "steady"
    streak: int | None = None
    album_diversity: float | None = None
    first_episode: str | None = None
    last_episode: str | None = None
    badge: str | None = None


class ArtistDetail(BaseModel):
    id: int | None = None
    artist: str
    plays: int
    episodes: int
    rli: float  # Bayesian Rollins Love Index
    coverage: float | None = None  # % of all episodes this artist appears in
    first_appearance: str | None = None
    last_appearance: str | None = None
    streak: int | None = None
    album_count: int | None = None
    album_diversity: float | None = None
    top_tracks: list["TrackCount"] = []
    album_breakdown: list["AlbumBreakdown"] = []
    timeline: list["TimelinePoint"] = []
    enrichment: Optional["ArtistEnrichment"] = None
    badge: str | None = None


class TrackPlay(BaseModel):
    broadcast: int | None = None
    date: str | None = None


class TrackCount(BaseModel):
    id: int | None = None
    title: str
    mbid: str | None = None
    plays: int
    last_played: str | None = None
    last_broadcast: int | None = None
    broadcasts: list[TrackPlay] = []
    episodes: int | None = None
    artist: str | None = None
    album: str | None = None


class AlbumBreakdown(BaseModel):
    album: str
    artist: str | None = None
    plays: int
    distinct_tracks: int
    artwork_url: str | None = None
    artwork_url_large: str | None = None
    release_date: str | None = None


class TimelinePoint(BaseModel):
    broadcast: int | None = None
    date: str
    plays: int


class ArtistEnrichment(BaseModel):
    mbid: str | None = None
    canonical_name: str | None = None
    country: str | None = None
    formed_year: int | None = None
    genres: list[str] | None = None
    tags: list[str] | None = None
    bio_summary: str | None = None
    lastfm_tags: list[str] | None = None
    lastfm_bio: str | None = None
    lastfm_listeners: int | None = None
    lastfm_playcount: int | None = None
    lastfm_url: str | None = None


# ── Album ──

class AlbumSummary(BaseModel):
    album: str
    artist: str
    plays: int
    distinct_tracks: int
    episodes: int
    last_played: str | None = None
    artwork_url: str | None = None


class ReleaseInfo(BaseModel):
    mbid: str | None = None
    title: str | None = None
    status: str | None = None
    date: str | None = None
    country: str | None = None
    format: str | None = None
    label: str | None = None
    track_count: int | None = None


class AlbumDetail(BaseModel):
    id: int | None = None
    artist_id: int | None = None
    album: str
    artist: str
    plays: int
    distinct_tracks: int
    total_tracks: int | None = None
    episodes: int
    tracks: list[TrackCount] = []
    timeline: list[TimelinePoint] = []
    heatmap: list[dict] = []
    artwork_url: str | None = None
    artwork_url_large: str | None = None
    mbid: str | None = None
    release_group_mbid: str | None = None
    release_date: str | None = None
    unplayed_tracks: list[str] = []
    releases: list[ReleaseInfo] = []


# ── Recommend / Bandcamp ──

class RecommendItem(BaseModel):
    url: str
    bandcamp_artist: str
    album_title: str
    episode_count: int
    first_seen: str | None = None
    last_seen: str | None = None
    episodes: list[int] = []


class RecommendArtist(BaseModel):
    bandcamp_artist: str
    unique_albums: int
    total_episodes: int
    last_seen: str | None = None


# ── Stats ──

class OverviewStats(BaseModel):
    episodes: int
    tracks: int
    unique_artists: int
    bandcamp_links: int
    date_range: tuple[str, str] | None = None


class TopItem(BaseModel):
    name: str
    value: int
    extra: dict = {}
